### PR TITLE
Recursion, Pt 2

### PR DESCRIPTION
**Pull request by mitschabaude**
* Follow-up to https://github.com/o1-labs/snarkyjs/pull/245
* RFC: https://github.com/o1-labs/snarkyjs/issues/89
* Adds new `Program` function to create a custom "proof system"
* Most of the logic for `Program` shares its implementation with `SmartContract`, even though `SmartContract` does not (yet?) use `Program` directly; common functionality was moved from `zkapp.ts` to `proof_system.ts`, and smart-contract-specific stuff was split out
* adds new example `src/examples/program.ts` which serves as rudimentary documentation for `Program`

### CODE REVIEW

File: src/lib/party.ts [hunk 1]
```diff
@@ -879,6 +879,14 @@ type PartiesProved = {
   memo: string;
 };
 
+/**
+ * The public input for zkApps consists of certain hashes of the transaction and of the proving Party which is constructed during method execution.
```
Comment (mitschabaude) [inline src/lib/party.ts:883]: this comment was just moved over from `zkapp.ts` where it made no sense anymore

File: src/lib/proof_system.ts [hunk 2]
```diff
@@ -41,3 +53,321 @@ let CompiledTag = {
     compiledTags.set(tag, compiledTag);
   },
 };
+
+function Program<
```
Comment (mitschabaude) [inline src/lib/proof_system.ts:57]: this is the new thing!

File: src/lib/zkapp.ts [hunk 3]
```diff
@@ -54,119 +53,69 @@ export function method<T extends SmartContract>(
...
+      // TODO: this needs to be done in a unified way for all parties that are created
+      assertPreconditionInvariants(this.self);
+      cleanPreconditionsCache(this.self);
+      return result;
     } else {
       // in a transaction, also add a lazy proof to the self party
       // (if there's no other authorization set)
       let auth = this.self.authorization;
       if (!('kind' in auth || 'proof' in auth || 'signature' in auth)) {
-        let previousProofs: Pickles.ProofWithPublicInput[] = [];
```
Comment (mitschabaude) [inline src/lib/zkapp.ts:108]: moved to `proof_system.ts` to be reused

**Review by mitschabaude — COMMENTED** [review]
some comments:

**Review by bkase — APPROVED** [review]

**Pull request merged**